api/layer/multipart_upload.go

```diff
@@ -201,35 +206,120 @@ func (n *layer) uploadPart(ctx context.Context, multipartInfo *data.MultipartInf
-		Container:    bktInfo.CID,
-		Creator:      bktInfo.Owner,
-		Attributes:   make([][2]string, 2),
-		Payload:      p.Reader,
-		CreationTime: TimeNow(ctx),
-		CopiesNumber: multipartInfo.CopiesNumber,
-	}
+	var (
+		bktInfo       = p.Info.Bkt
+		payloadReader = p.Reader
+		decSize       = p.Size
+		attributes    [][2]string
+	)
 
-	decSize := p.Size
 	if p.Info.Encryption.Enabled() {
 		r, encSize, err := encryptionReader(p.Reader, uint64(p.Size), p.Info.Encryption.Key())
 		if err != nil {
 			return nil, fmt.Errorf("failed to create ecnrypted reader: %w", err)
 		}
-		prm.Attributes = append(prm.Attributes, [2]string{AttributeDecryptedSize, strconv.FormatInt(p.Size, 10)})
-		prm.Payload = r
+		attributes = append(attributes, [2]string{AttributeDecryptedSize, strconv.FormatInt(p.Size, 10)})
+		payloadReader = r
 		p.Size = int64(encSize)
 	}
 
-	prm.Attributes[0][0], prm.Attributes[0][1] = UploadIDAttributeName, p.Info.UploadID
-	prm.Attributes[1][0], prm.Attributes[1][1] = UploadPartNumberAttributeName, strconv.Itoa(p.PartNumber)
+	var (
+		splitPreviousID      oid.ID
+		isSetSplitPreviousID bool
+		multipartHash        = sha256.New()
+		tzHash               hash.Hash
+	)
 
-	id, hash, err := n.objectPutAndHash(ctx, prm, bktInfo)
+	if n.neoFS.IsHomomorphicEnabled() {
+		tzHash = tz.New()
+	}
+
+	lastPart, err := n.treeService.GetLastPart(ctx, bktInfo, multipartInfo.ID)
 	if err != nil {
-		return nil, err
+		// if ErrPartListIsEmpty, there is the first part of multipart.
+		if !errors.Is(err, ErrPartListIsEmpty) {
+			return nil, fmt.Errorf("getLastPart: %w", err)
+		}
+	} else {
+		// try to restore hash state from the last part.
+		if binaryUnmarshaler, ok := multipartHash.(encoding.BinaryUnmarshaler); ok {
+			if err = binaryUnmarshaler.UnmarshalBinary(lastPart.MultipartHash); err != nil {
+				return nil, fmt.Errorf("unmarshalBinary: %w", err)
+			}
+		}
+
+		if n.neoFS.IsHomomorphicEnabled() {
+			if binaryUnmarshaler, ok := tzHash.(encoding.BinaryUnmarshaler); ok {
+				if err = binaryUnmarshaler.UnmarshalBinary(lastPart.HomoHash); err != nil {
+					return nil, fmt.Errorf("unmarshalBinary: %w", err)
+				}
+			}
+		}
+
+		isSetSplitPreviousID = true
+		splitPreviousID = lastPart.OID
 	}
 
+	var (
+		id           oid.ID
+		elements     []string
+		creationTime = TimeNow(ctx)
+		// User may upload part large maxObjectSize in NeoFS. From users point of view it is a single object.
+		// We have to calculate the hash from this object separately.
+		currentPartHash = sha256.New()
+	)
+
+	multipartHashes := []hash.Hash{multipartHash, currentPartHash}
+	if tzHash != nil {
+		multipartHashes = append(multipartHashes, tzHash)
+	}
+
+	prm := PrmObjectCreate{
+		Container:       bktInfo.CID,
+		Creator:         bktInfo.Owner,
+		Attributes:      attributes,
+		CreationTime:    creationTime,
+		CopiesNumber:    multipartInfo.CopiesNumber,
+		SplitID:         multipartInfo.SplitID,
+		MultipartHashes: multipartHashes,
+	}
+
+	dt := n.buffers.Get()
```

Comment:
not critical but dt?.. data? destination?

Reply:
data already occupied for this place, joined with the next line